core/pta/attestation.c

```diff
@@ -0,0 +1,212 @@
+{
+	struct fobj *fobj = r->mobj->ops->get_fobj(r->mobj);
+	struct file *f = to_file_may_fail(r->mobj);
+	uint64_t offs_and_size[2] = { };
+	TEE_Result res = TEE_SUCCESS;
+	unsigned int poffs = 0;
+	uint8_t *tag = NULL;
+
+	/* Regions we're interested in should all have a struct file */
+	assert(f);
+
+	tag = file_get_tag(f);
+	assert(tag);
+
+	res = crypto_hash_update(ctx, tag, FILE_TAG_SIZE);
+	if (!res)
+		return res;
+
+	file_lock(f);
+	res = file_find_page_offset(f, fobj, &poffs);
+	file_unlock(f);
+	if (!res)
+		return res;
+
+	offs_and_size[0] = poffs * SMALL_PAGE_SIZE;
+	offs_and_size[1] = r->mobj->size;
+
+	return crypto_hash_update(ctx, (uint8_t *)offs_and_size,
+				  sizeof(offs_and_size));
+}
+
+static TEE_Result hash_update(void *ctx, struct vm_region *r, uint32_t method)
+{
+	switch (method) {
+	case PTA_ATTESTATION_HASH_METHOD_FULL:
+		return crypto_hash_update(ctx, (uint8_t *)r->va, r->size);
+	case PTA_ATTESTATION_HASH_METHOD_TAGS:
+		return hash_update_method_tags(ctx, r);
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+}
+
+static TEE_Result hash_regions(struct vm_info *vm_info, uint8_t *hash,
+			       uint32_t method)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct vm_region *r = NULL;
+	struct vm_region **regions = NULL;
+	size_t nregions = 0;
+	void *ctx = NULL;
+	size_t i = 0;
+
+	res = crypto_hash_alloc_ctx(&ctx, TEE_ALG_SHA256);
+	if (res)
+		return res;
+
+	res = crypto_hash_init(ctx);
+	if (res)
+		goto out;
+
+	/*
+	 * Make an array of region pointers so we can use qsort() to order it.
+	 */
+
+	TAILQ_FOREACH(r, &vm_info->regions, link)
+		if (is_region_valid(r))
+			nregions++;
+
+	regions = malloc(nregions * sizeof(*regions));
+	if (!regions) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	TAILQ_FOREACH(r, &vm_info->regions, link)
+		if (is_region_valid(r))
+			regions[i++] = r;
+
+	/*
+	 * Sort regions so that they are in a consistent order even when TA ASLR
+	 * is enabled.
+	 */
+	qsort(regions, nregions, sizeof(*regions), cmp_regions);
+
+	/* Hash regions in order */
+	for (i = 0; i < nregions; i++) {
+		r = regions[i];
+		DMSG("va %p size %zu", (void *)r->va, r->size);
+		res = hash_update(ctx, r, method);
+		if (res)
+			goto out;
+	}
+
+	res = crypto_hash_final(ctx, hash, TEE_SHA256_HASH_SIZE);
+out:
+	free(regions);
+	crypto_hash_free_ctx(ctx);
+	return res;
+}
+
+static TEE_Result hash_ta(struct user_mode_ctx *uctx, uint32_t types,
+			  TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t method = params[0].value.a;
+	uint8_t *hash = params[1].memref.buffer;
+	size_t hash_sz = params[1].memref.size;
+	TEE_Result res = TEE_SUCCESS;
+	struct ts_session *s = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE, TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!hash || hash_sz != TEE_SHA256_HASH_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	s = ts_pop_current_session();
```

Comment:
This shouldn't be needed.

Reply:
Cargo-culted from https://github.com/OP-TEE/optee_os/blob/3.15.0/core/pta/system.c#L226-L228 :wink: 
Just to clarify the reason why it is needed there and not here: we're not switching to user mode here, right?

Will fix.

Reply:
When you activate a session you also activates the mappings of that context. In the case of PTAs they always borrow the mappings from the calling session so the PTA can access memory in the calling user TA.

Reply:
Got it, thanks.